File: backend/app/modules/congregations/geo.py
"""Geographic constants for congregation addresses.

Countries are stored as ISO 3166-1 alpha-2 codes ("PL", "DE") so the UI can
render a localized name via `Intl.DisplayNames` without translation files.

Provinces are stored as ASCII slugs. Only Poland has a defined subdivision
list today; addresses in other countries keep `province = NULL`.
"""

DEFAULT_COUNTRY = "PL"

COUNTRY_CODE_PATTERN = r"^[A-Z]{2}$"

# ISO 3166-2:PL — the 16 Polish voivodeships, as ASCII slugs.
POLISH_PROVINCES: tuple[str, ...] = (
    "dolnoslaskie",
    "kujawsko-pomorskie",
    "lubelskie",
    "lubuskie",
    "lodzkie",
    "malopolskie",
    "mazowieckie",
    "opolskie",
    "podkarpackie",
    "podlaskie",
    "pomorskie",
    "slaskie",
    "swietokrzyskie",
    "warminsko-mazurskie",
    "wielkopolskie",
    "zachodniopomorskie",
)

PROVINCES_BY_COUNTRY: dict[str, tuple[str, ...]] = {
    "PL": POLISH_PROVINCES,
}


def is_valid_province(country: str, province: str | None) -> bool:
    """A province must belong to its country's subdivision list, when one exists."""
    if province is None:
        return True
    known = PROVINCES_BY_COUNTRY.get(country)
    if known is None:
        return False
    return province in known

File: src/modules/congregations/utils/geo.ts
/**
 * Countries are stored as ISO 3166-1 alpha-2 codes and rendered with
 * `Intl.DisplayNames`, so a new country needs no translation entries.
 */

/** Countries offered when editing an address. Extend as congregations appear. */
export const COUNTRY_CODES: readonly string[] = [
  'PL',
  'DE',
  'AT',
  'BE',
  'BY',
  'CZ',
  'DK',
  'ES',
  'FR',
  'GB',
  'IE',
  'IS',
  'IT',
  'LT',
  'LV',
  'NL',
  'NO',
  'SE',
  'SK',
  'UA',
  'US',
  'CA',
]

export const DEFAULT_COUNTRY_CODE = 'PL'

/** ISO 3166-2:PL — voivodeship slugs, matching `backend/app/modules/congregations/geo.py`. */
export const POLISH_PROVINCES: readonly string[] = [
  'dolnoslaskie',
  'kujawsko-pomorskie',
  'lubelskie',
  'lubuskie',
  'lodzkie',
  'malopolskie',
  'mazowieckie',
  'opolskie',
  'podkarpackie',
  'podlaskie',
  'pomorskie',
  'slaskie',
  'swietokrzyskie',
  'warminsko-mazurskie',
  'wielkopolskie',
  'zachodniopomorskie',
]

const PROVINCES_BY_COUNTRY: Record<string, readonly string[]> = {
  PL: POLISH_PROVINCES,
}

/** Voivodeship slugs carry no diacritics; restore them for display. */
const POLISH_PROVINCE_LABELS: Record<string, string> = {
  'dolnoslaskie': 'dolnośląskie',
  'kujawsko-pomorskie': 'kujawsko-pomorskie',
  'lubelskie': 'lubelskie',
  'lubuskie': 'lubuskie',
  'lodzkie': 'łódzkie',
  'malopolskie': 'małopolskie',
  'mazowieckie': 'mazowieckie',
  'opolskie': 'opolskie',
  'podkarpackie': 'podkarpackie',
  'podlaskie': 'podlaskie',
  'pomorskie': 'pomorskie',
  'slaskie': 'śląskie',
  'swietokrzyskie': 'świętokrzyskie',
  'warminsko-mazurskie': 'warmińsko-mazurskie',
  'wielkopolskie': 'wielkopolskie',
  'zachodniopomorskie': 'zachodniopomorskie',
}

export function provincesForCountry(countryCode: string): readonly string[] {
  return PROVINCES_BY_COUNTRY[countryCode] ?? []
}

export function provinceLabel(province: string): string {
  return POLISH_PROVINCE_LABELS[province] ?? province
}

/**
 * Localized country name, e.g. 'PL' -> 'Polska' (pl) / 'Poland' (en).
 * Falls back to the raw code where `Intl.DisplayNames` is unavailable.
 */
export function countryLabel(countryCode: string, locale: string): string {
  try {
    const displayNames = new Intl.DisplayNames([locale], { type: 'region' })
    return displayNames.of(countryCode) ?? countryCode
  } catch {
    return countryCode
  }
}

export interface ICountryOption {
  code: string
  label: string
}

/** Country options sorted by localized name, with the default country first. */
export function countryOptions(locale: string): ICountryOption[] {
  const collator = new Intl.Collator(locale)
  return COUNTRY_CODES.map<ICountryOption>((code) => ({ code, label: countryLabel(code, locale) }))
    .sort((a, b) => {
      if (a.code === DEFAULT_COUNTRY_CODE) return -1
      if (b.code === DEFAULT_COUNTRY_CODE) return 1
      return collator.compare(a.label, b.label)
    })
}
